gwcs/wcs.py

```diff
@@ -634,3 +641,235 @@ def fix_inputs(self, fixed):
+        # The fitting section.
+        # Fit the forward case.
+        fit_poly_x, fit_poly_y, max_resid, rms = fit_2D_poly(degree, max_error, max_rms,
+                                                             u, v, undist_x, undist_y,
+                                                             verbose=verbose)
+        cdmat = np.array([[fit_poly_x.c1_0.value, fit_poly_x.c0_1.value],
+                          [fit_poly_y.c1_0.value, fit_poly_y.c0_1.value]])
+        det = cdmat[0, 0] * cdmat[1, 1] - cdmat[0, 1] * cdmat[1, 0]
+        U = (cdmat[1, 1] * undist_x - cdmat[0, 1] * undist_y) / det
+        V = (-cdmat[1, 0] * undist_x + cdmat[0, 0] * undist_y) / det
+        fit_inv_poly_u, fit_inv_poly_v, max_resid, rms = fit_2D_poly(degree,
+                                                            max_inv_error, max_inv_rms,
+                                                            U, V, u - U, v - V,
+                                                            verbose=verbose)
+        pdegree = fit_poly_x.degree
+        hdr['a_order'] = pdegree
+        hdr['b_order'] = pdegree
+        ipdegree = fit_inv_poly_u.degree
+        hdr['ap_order'] = ipdegree
+        hdr['bp_order'] = ipdegree
+        cd, sip_poly_x, sip_poly_y = reform_poly_coefficients(fit_poly_x, fit_poly_y)
+        store_2D_coefficients(hdr, sip_poly_x, 'A')
+        store_2D_coefficients(hdr, sip_poly_y, 'B')
+        store_2D_coefficients(hdr, fit_inv_poly_u, 'AP', keeplinear=True)
+        store_2D_coefficients(hdr, fit_inv_poly_v, 'BP', keeplinear=True)
+        hdr['cd1_1'] = cd[0][0]
+        hdr['cd1_2'] = cd[0][1]
+        hdr['cd2_1'] = cd[1][0]
+        hdr['cd2_2'] = cd[1][1]
+        hdr['sipmxerr'] = max_resid
+        hdr['siprms'] = rms
+        # # for test purposes, make sip equivalent in GWCS
+        # cdmat = np.array([[cd[0][0], cd[0][1]], [cd[1][0], cd[1][1]]])
+        # # cdmat = [[cd[0][0], cd[1][0]], [cd[0][1], cd[1][1]]]
+        # sip_poly_test_x = sip_poly_x.copy()
+        # sip_poly_test_y = sip_poly_y.copy()
+        # # sip_poly_test_x.c1_0 = 1
+        # # sip_poly_test_y.c0_1 = 1
+        # shift = (Shift(-crpix1) & Shift(-crpix2))
+        # aff = AffineTransformation2D(cdmat, [0,0])
+
+        # transform = (shift | Mapping([0,1,0,1]) |
+        #             (Mapping([0], n_inputs=2) + sip_poly_test_x) &
+        #             (Mapping([1], n_inputs=2) + sip_poly_test_y) |
+        #             aff |
+        #             Pix2Sky_TAN() | RotateNative2Celestial(crval1, crval2, 180))
+        # transform2 = ((Shift(-crpix1) & Shift(-crpix2)) | Mapping([0,1,0,1]) |
+        #               (fit_poly_x & fit_poly_y) | Pix2Sky_TAN() |
+        #               RotateNative2Celestial(crval1, crval2, 180))
+        return hdr
+
+
+def fit_2D_poly(degree, max_error, max_rms, u, v, x, y, verbose=False):
+    """
+    Fit a pair of ordinary 2D polynomial to the supplied inputs (u, v) and
+    outputs (x, y)
```

Comment:
Documentation should include function parameters, return value and maybe what kind of exception this function might throw and under what circumstances.

Reply:
Added some of that, but review again.

Reply:
Also, do you intend to have `fit_2D_poly()`, `reform_poly_coefficients()`, `compute_distance_residual ()`, and `store_2D_coefficients()` functions exposed to the user? Or should these be private? If so, please consider prepending `_`.

Reply:
Good point.

Reply:
> Added some of that, but review again.

I don't see it.